{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4abfe648",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import matplotlib.pyplot as plt\n",
    "import numpy as np\n",
    "import seaborn as sns\n",
    "import io\n",
    "import os"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "219fef82",
   "metadata": {},
   "outputs": [],
   "source": [
    "\n",
    "def read_vcf(path):\n",
    "    '''\n",
    "    input: path to vcf file\n",
    "    returns: pandas df from vcf, the header is ignored\n",
    "    #partly taken from https://gist.github.com/dceoy/99d976a2c01e7f0ba1c813778f9db744, 20.12.2021\n",
    "    '''\n",
    "    with open(path, 'r') as f: \n",
    "        lines = [l for l in f if not l.startswith('##')]\n",
    "        \n",
    "    return pd.read_csv(\n",
    "        io.StringIO(''.join(lines)),\n",
    "        dtype={'#CHROM': str, 'POS': int, 'ID': str, 'REF': str, 'ALT': str,\n",
    "               'QUAL': float, 'FILTER': str, 'INFO': str},\n",
    "        sep='\\t'\n",
    "    ).rename(columns={'#CHROM': 'CHROM'})"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "3f957033",
   "metadata": {},
   "outputs": [],
   "source": [
    "def read_INFO(df):\n",
    "    '''\n",
    "    input: pandas vcf dataframe\n",
    "    returns: dataframe containing recoded column INFO from the vcf dataframe\n",
    "    '''\n",
    "    new = df[\"INFO\"].str.split(\";\", expand = True)\n",
    "    new = new.dropna()\n",
    "    new.columns = pd.DataFrame(new.iloc[0,].str.split(\"=\")).apply(lambda row : row[0][0], axis = 1)\n",
    "    new_new = new.applymap(lambda cell: cell.split(\"=\")[-1])\n",
    "    return(new_new)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "957959b7",
   "metadata": {},
   "outputs": [],
   "source": [
    "def number_of_heterozygots(df, column):\n",
    "    '''\n",
    "    input: pandas vcf dataframe, name of the sample\n",
    "    returns: valuecounts of mono/heterozygots (0/0,0/1 ...)\n",
    "    '''\n",
    "    \n",
    "    new = df[column].str.split(\":\", expand = True)[0]\n",
    "    new = new.dropna()\n",
    "    return(new.value_counts())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "b4b41f22",
   "metadata": {},
   "outputs": [],
   "source": [
    "def numeric_INFO(INFO):\n",
    "    \n",
    "    INFO['MQ'] = pd.to_numeric(INFO['MQ'])\n",
    "    INFO['AN'] = pd.to_numeric(INFO['AN'])\n",
    "    INFO['BaseQRankSum'] = pd.to_numeric(INFO['BaseQRankSum'])\n",
    "    INFO['ClippingRankSum'] = pd.to_numeric(INFO['ClippingRankSum'])\n",
    "    INFO['ExcessHet'] = pd.to_numeric(INFO['ExcessHet'])\n",
    "    INFO['FS'] = pd.to_numeric(INFO['FS'])\n",
    "    INFO['MQRankSum'] = pd.to_numeric(INFO['MQRankSum'])\n",
    "    INFO['QD'] = pd.to_numeric(INFO['QD'])\n",
    "    INFO['ReadPosRankSum'] = pd.to_numeric(INFO['ReadPosRankSum'])\n",
    "    INFO['SOR'] = pd.to_numeric(INFO['SOR'])\n",
    "    "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "ad4aecab",
   "metadata": {},
   "outputs": [],
   "source": [
    "def plot_dist(df, column, threshold):\n",
    "    '''\n",
    "    Plot the distribution of the metric with the treshold\n",
    "    Default thresholds GATK hard filtering (GATK 4.0 VariantFiltration):\n",
    "        For indels:\n",
    "            QD < 2.0\n",
    "            QUAL < 30.0\n",
    "            FS > 200.0\n",
    "            ReadPosRankSum < -20.0\n",
    "            \n",
    "        For SNPs:\n",
    "            QD < 2.0\n",
    "            QUAL < 30.0\n",
    "            SOR > 3.0\n",
    "            FS > 60.0\n",
    "            MQ < 40.0\n",
    "            MQRankSum < -12.5\n",
    "            ReadPosRankSum < -8.0\n",
    "\n",
    "    '''\n",
    "    \n",
    "    sns.displot(df[column], x=column, kind=\"kde\")\n",
    "    plt.axvline(treshold, 0, 1, color = 'red')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "24674201",
   "metadata": {},
   "outputs": [],
   "source": [
    "def change_vcf_sample_names(old_vcf_path, new_vcf_path, new_names):\n",
    "    \"\"\"\n",
    "    Change the names of samples in vcf: last columns are renamed to the ones given in the list\n",
    "    input: \n",
    "        old_vcf - path to the vcf file, which samples need to be renamed\n",
    "        new_vcf - path to the new vcf file with renamed samples\n",
    "        new_names - path to the file with new names separated by \"\\t\"\n",
    "        \n",
    "    return: old name samples\n",
    "    \n",
    "    ! UPD: similar thing can be done in shell by \"bcftools reheader -s new_names.list -o new_vcf_path old_vcf_path\"\n",
    "    \n",
    "    \"\"\"\n",
    "    old_names = []\n",
    "    with open(new_names, 'r') as sample_names:\n",
    "        samples = [l.strip() for l in sample_names]\n",
    "        if (len(samples)== 0):\n",
    "            print('FILE WITH NEW NAMES IS EMPTY!')\n",
    "    with open(old_vcf_path, 'r') as f:\n",
    "        with open(new_vcf_path, 'w') as new_vcf: \n",
    "            for l in f: \n",
    "                if l.startswith('#CHROM'):\n",
    "                    headers = l.split('\\t')\n",
    "                    if (headers[-(len(samples)+1)]!=\"FORMAT\"):\n",
    "                        print('MAKE SURE THAT NUMBER OF NAMES IS EQUAL TO NUMBER OF SAMPLES IN THE VCF!')\n",
    "                    old_names = headers[-len(samples):] \n",
    "                    del headers[-len(samples):]\n",
    "                    headers+=samples\n",
    "                    print(*headers, sep = \"\\t\", file = new_vcf) \n",
    "                else:\n",
    "                    print(l, file = new_vcf)\n",
    "    return(old_names)             "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "75a193fa",
   "metadata": {},
   "outputs": [],
   "source": [
    "def clean_vcf(path_in, path_out):\n",
    "    '''\n",
    "    Deleats all the raws that did not pass one of the filters\n",
    "    input: path to input vcf\n",
    "    output: path to output cleaned vcf\n",
    "    '''  \n",
    "    with open(path_in, 'r') as f:\n",
    "        comments = []\n",
    "        raws = []\n",
    "        for l in f:\n",
    "            if l.startswith('##'):\n",
    "                comments.append(l)\n",
    "            else:\n",
    "                raws.append(l)\n",
    "\n",
    "    df = pd.read_csv(io.StringIO(''.join(raws)),\n",
    "                     dtype={'#CHROM': str, 'POS': int, 'ID': str, 'REF': str, 'ALT': str,\n",
    "                            'QUAL': float, 'FILTER': str, 'INFO': str}, sep='\\t').rename(columns={'#CHROM': 'CHROM'})\n",
    "    df_cleaned = df[df['FILTER']=='PASS']\n",
    "    with open(path_out, 'w') as output:\n",
    "        print(*comments, sep = '', file = output)\n",
    "    with open(path_out, 'a') as output:    \n",
    "        df_cleaned.to_csv(output, sep='\\t',index=False, header=True)\n",
    "    return df_cleaned    "
   ]
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.8.8"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
